compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+        """Index all tuples of min_match_length in self.window[:end_pos]
+        into the dict. The last tuple to be indexed will start at end_pos-min_match_length+1
+        This uses self.window_indexed_till to ensure we do not reindex things already indexed.
+
+        Args:
+            end_pos (int): end position in window to index
+        """
+        for end_pos_substr in range(self.window_indexed_till, end_pos + 1):
+            start_pos_substr = end_pos_substr - self.min_match_length
+            if start_pos_substr < 0:
+                continue
+            substr = tuple(self.window[start_pos_substr:end_pos_substr])
+            self.insert_substring_into_dict(substr, start_pos_substr)
+        self.window_indexed_till = end_pos + 1
+
+    def encode_lz77_sequences(self, lz77_sequences: LZ77Sequence):
+        """Perform entropy encoding of the LZ77 sequences and return the encoded bitarray.
+
+        We simply use the Elias Delta code for encoding all three, but first we preprocess a bit to
+        optimize the encoding.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        literal_counts = [l.literal_count for l in lz77_sequences]
+        # subtract min_match_length from match_length
+        match_lengths_processed = [l.match_length - self.min_match_length for l in lz77_sequences]
+        # subtract 1 from match offset because it's at least 1
+        match_offsets_processed = [l.match_offset - 1 for l in lz77_sequences]
+        # Combine all streams into one and then apply Elias Delta.
+        # first encode the min_match_length (needed because we subtract it from match_lengths)
+        combined = (
+            [self.min_match_length]
+            + literal_counts
+            + match_lengths_processed
+            + match_offsets_processed
+        )
+        combined_encoding = EliasDeltaUintEncoder().encode_block(DataBlock(combined))
+        return (
+            uint_to_bitarray(len(combined_encoding), ENCODED_BLOCK_SIZE_HEADER_BITS)
+            + combined_encoding
+        )
+
+    def encode_literals(self, literals: List):
+        """Perform entropy encoding of the literals and return the encoded bitarray.
+
+        We apply Huffman coding for the literals and also store the counts to enable the
+        decoder to construct the same Huffman tree.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        # first encode the literals with empirical Huffman code
+        counts = DataBlock(literals).get_counts()
+        if len(counts) > 0:
+            prob_dict = ProbabilityDist.normalize_prob_dict(counts).prob_dict
+            # let's sort prob_dist by the alphabet to make sure we get exact same ordering
+            # during decompression as well, which makes sure it's the same Huffman tree!
+            prob_dist_sorted = ProbabilityDist({i: prob_dict[i] for i in sorted(prob_dict)})
```

Comment:
should we add this in the Huffman code construction, so that the Huffman coder has less issues?

Reply:
I think it's better to make it the caller's responsibility and perhaps adding a note in the Huffman coder. In retrospect this seems a very obvious thing, I think the confusion is caused by Python's dict maybe (if we were passing a list I feel this would be clearer).

A couple reasons to not add this to Huffman:
- might cause trouble if the type is not sortable
- this issue is not specific to Huffman, it applies also to arithmetic coder, range coder etc.

Reply:
sounds good, yeah we should make the prob_dist an orderedDict more explicitly (`dict` is ordered nowadays, but is not very explicit)